### PR TITLE
Format Rules and Support for Sqlserver Select Clause

### PR DESCRIPTION
**Pull request by ThanoshanMV**
For #6478 

Changes proposed in this pull request:
- Format and align rules according to the [documentation](https://docs.microsoft.com/en-us/sql/t-sql/queries/select-transact-sql?view=sql-server-ver15).
- Support entirely for SQLServer [SELECT Clause](https://docs.microsoft.com/en-us/sql/t-sql/queries/select-clause-transact-sql?view=sql-server-ver15).
- The SQLServer TOP clause specifies the first set of rows to be returned from the query result set. Its grammar is `[ TOP ( expression ) [ PERCENT ] [ WITH TIES ] ]`.
- However, the grammar of the TOP clause in the project was: `TOP LP_? topNum RP_? PERCENT? (WITH TIES)? (ROW_NUMBER LP_ RP_ OVER LP_ orderByClause RP_)?`. So, TOP clause covered [ROW_NUMBER](https://docs.microsoft.com/en-us/sql/t-sql/functions/row-number-transact-sql?view=sql-server-ver15) function from the projection.
- After proofreading of TOP clause, I've added `expression-projection` element inside `projections` element to describe the `ROW_NUMBER()` function.
- Parse tree of a SQL which contains a TOP clause after proofreading:
top-clause-parse-tree
- Please let me know if there are any changes I need to make.

### CODE REVIEW

**Comment by ThanoshanMV** [comment]
Hi @strongduanmu, I had test failures in the shardingsphere-rewrite-test. I changed some expected values in the shardingsphere-rewrite-test. I read about [SQL Rewrite Test](https://shardingsphere.apache.org/document/current/en/reference/test/module-test/rewrite-test/) and other [references](https://shardingsphere.apache.org/document/current/en/reference/sharding/rewrite/#column-derivation). However, I don't understand why test failures occur and how to resolve them. Can you please guide me regarding this?

**Comment by strongduanmu** [comment]
> 
@ThanoshanMV Thank you for your contribution, I will take a look at this pr.

**Comment by strongduanmu** [comment]
> * However, the grammar of the TOP clause in the project was: `TOP LP_? topNum RP_? PERCENT? (WITH TIES)? (ROW_NUMBER LP_ RP_ OVER LP_ orderByClause RP_)?`. So, TOP clause covered [ROW_NUMBER](https://docs.microsoft.com/en-us/sql/t-sql/functions/row-number-transact-sql?view=sql-server-ver15) function from the projection.

Yes, I think we should split top and row_number to different grammar rule.

File: shardingsphere-test/shardingsphere-rewrite-test/src/test/resources/scenario/sharding/case/select.xml [hunk 1]
```diff
@@ -415,26 +415,26 @@
 
     <rewrite-assertion id="select_top_with_multiple_route_for_parameters_for_sqlserver" db-types="SQLServer">
         <input sql="SELECT * FROM (SELECT TOP(?) row_number() OVER (ORDER BY o.account_id) AS rownum_, o.account_id FROM t_account o WHERE o.account_id IN (100, 101)) AS row_ WHERE row_.rownum_ &gt; ?" parameters="110, 100" />
-        <output sql="SELECT * FROM (SELECT TOP(?) row_number() OVER (ORDER BY o.account_id) AS rownum_, o.account_id FROM t_account_0 o WHERE o.account_id IN (100, 101)) AS row_ WHERE row_.rownum_ &gt; ?" parameters="110, 0" />
-        <output sql="SELECT * FROM (SELECT TOP(?) row_number() OVER (ORDER BY o.account_id) AS rownum_, o.account_id FROM t_account_1 o WHERE o.account_id IN (100, 101)) AS row_ WHERE row_.rownum_ &gt; ?" parameters="110, 0" />
+        <output sql="SELECT * FROM (SELECT TOP(?) row_number() OVER (ORDER BY o.account_id) AS rownum_, o.account_id FROM t_account_0 o WHERE o.account_id IN (100, 101)) AS row_ WHERE row_.rownum_ &gt; ?" parameters="110, 100" />
```
Comment (strongduanmu) [inline shardingsphere-test/shardingsphere-rewrite-test/src/test/resources/scenario/sharding/case/select.xml:418]: Hi @ThanoshanMV, the initial test case is correct. When we perform distributed paging query on multiple data nodes, we need to obtain the data starting from the 0 index, and then process it in the ShardingSphere result merge engine. 

I think a better way to solve the problem is to deploy an instance of SQLServer locally via docker, then create a sharding table and insert some data, and then execute paging query sql to investigate the cause of the problem. You can focus on investigating the `ShardingPaginationParameterRewriter` class.

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: shardingsphere-test/shardingsphere-rewrite-test/src/test/resources/scenario/sharding/case/select.xml (hunk 1 above)
Comment (strongduanmu) [reply shardingsphere-test/shardingsphere-rewrite-test/src/test/resources/scenario/sharding/case/select.xml:418]: In addition, the final result is processed through the TopAndRowNumberDecoratorMergedResult class, you can also refer to it.

Comment (ThanoshanMV) [reply shardingsphere-test/shardingsphere-rewrite-test/src/test/resources/scenario/sharding/case/select.xml:418]: Okay @strongduanmu, I'll check them.

**Comment by ThanoshanMV** [comment]
I'll continue to work on this PR after the GSoC competition finishes and update the progress here. After fixing this issue, I'll reopen this PR. If there are any issues with reopening this PR, I'll make a new PR.

**Pull request closed without merging**